frontend/components/instruments/SliderInstrument.js

```diff
@@ -1,46 +1,137 @@
-import React, {useState} from "react";
+import React, {useEffect, useState, useRef} from "react";
 import PropTypes from "prop-types";
 import Slider from "../inputs/Slider";
 
-const SliderInstrument = ({samples}) => {
-    if (samples === null) { return ('Waiting to load data to create an instrument.'); }
+/*
+ * Takes base64-encoded wav files and converts them to an ArrayBuffer
+ * of the decoded data, which what AudioBufferSourceNode needs to consume
+ */
+const sampleBase64StrToArrayBuffer = (sampleStr) => {
+    const byte_str = atob(sampleStr);
+    const buffer = new ArrayBuffer(byte_str.length * 2);
+    const buffer_view = new Uint8Array(buffer);
+    for (let i = 0; i < byte_str.length; i++) {
+        buffer_view[i] = byte_str.charCodeAt(i);
+    }
+    return buffer;
+};
+
+
+/*
+ * An audio playback component that plays a single sample using the Web Audio API
+ */
+const SamplePlayer = ({
+    isPlaying,
+    loop,
+    sample,
+    volume,
+}) => {
+    // TODO(ra): probably we want the AudioContext to live in the container component,
+    //           so we can apply compression to the whole thing.
+    const audioContextRef = useRef(new AudioContext());
+    const audioSourceRef = useRef(null);
+    const audioGainNodeRef = useRef(audioContextRef.current.createGain());
+    const previousPlayState = useRef(isPlaying);
+    const [shouldSetupAudio, setShouldSetupAudio] = useState(true);
+
+    const setupAudioRouting = () => {
+        const audioContext = audioContextRef.current;
 
-    const [isPlaying, setIsPlaying] = useState(true);
+        const audioGainNode = audioGainNodeRef.current;
+        audioGainNode.connect(audioContext.destination);
 
-    const adjustAudioVolume = (audioElementId, newValue) => {
-        const audioTag = document.getElementById(audioElementId);
-        audioTag.volume = newValue / 100;
+        // NOTE(ra): Could do this via createMediaElementSource instead - maybe simpler. Not sure.
+        const audioSource = audioContext.createBufferSource();
+        audioSourceRef.current = audioSource;
+        audioSource.loop = loop;
+        audioSource.connect(audioGainNode);
+
+        // AudioBufferSourceNode is a single-use object: once the buffer plays once, we have to
+        // throw it away and do this audio setup again.
+        audioSource.onended = () => setShouldSetupAudio(true);
+
+        const audioBuffer = sampleBase64StrToArrayBuffer(sample);
```

Comment:
This data transformation should not happen whenever playback resumes: we should cache the result in the container component or in a ref. I'll do in a next step.